contracts/src/base/FROSTCoordinator.sol

```diff
@@ -0,0 +1,108 @@
+// SPDX-License-Identifier: GPL-3.0-only
+pragma solidity ^0.8.30;
+
+import {FROSTMerkleMap} from "@/lib/FROSTMerkleMap.sol";
+import {Secp256k1} from "@/lib/Secp256k1.sol";
+
+/// @title FROST Coordinator
+/// @notice An onchain coordinator for FROST key generation and signing.
+contract FROSTCoordinator {
+    using FROSTMerkleMap for FROSTMerkleMap.T;
+
+    type GroupId is bytes32;
+
+    struct GroupParameters {
+        uint128 count;
+        uint128 threshold;
+    }
+
+    struct Group {
+        FROSTMerkleMap.T participants;
+        GroupParameters parameters;
+        Secp256k1.Point key;
+    }
+
+    struct KeyGenCommitment {
+        Secp256k1.Point[] c;
+        Secp256k1.Point r;
+        uint256 mu;
+    }
+
+    struct KeyGenSecretShare {
+        Secp256k1.Point y;
+        uint256[] f;
+    }
+
+    event KeyGen(GroupId indexed id, bytes32 participants, uint128 count, uint128 threshold);
+    event KeyGenAborted(GroupId indexed id);
+    event KeyGenCommitted(GroupId indexed id, uint256 index, KeyGenCommitment commitment);
+    event KeyGenSecretShared(GroupId indexed id, uint256 index, KeyGenSecretShare share);
+
+    error InvalidGroupParameters();
+    error InvalidKeyGenCommitment();
+    error InvalidKeyGenSecretShare();
+
+    // forge-lint: disable-next-line(mixed-case-variable)
+    mapping(GroupId => Group) private $groups;
+
+    /// @notice Initiate a distributed key generation ceremony.
+    function keygen(uint96 nonce, bytes32 participants, uint128 count, uint128 threshold)
```

Comment:
nit: when reading the code I misinterpret `participants` as a list or array, not as a (root) hash